File: src/main/java/no/fint/provider/springer/model/FasttilleggRepository.java
package no.fint.provider.springer.model;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import no.fint.event.model.Event;
import no.fint.event.model.ResponseStatus;
import no.fint.event.model.Status;
import no.fint.model.administrasjon.personal.PersonalActions;
import no.fint.model.resource.FintLinks;
import no.fint.model.resource.administrasjon.personal.FasttilleggResource;
import no.fint.provider.springer.behaviour.Behaviour;
import no.fint.provider.springer.service.Handler;
import no.fint.provider.springer.service.IdentifikatorFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.Resource;
import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
import org.springframework.stereotype.Repository;

import javax.annotation.PostConstruct;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Slf4j
@Repository
public class FasttilleggRepository implements Handler {

    @Autowired
    ObjectMapper objectMapper;

    @Autowired
    IdentifikatorFactory identifikatorFactory;

    @Autowired
    List<Behaviour<FasttilleggResource>> behaviours;

    private Collection<FasttilleggResource> repository = new ConcurrentLinkedQueue<>();

    @PostConstruct
    public void init() throws IOException {
        for (Resource r : new PathMatchingResourcePatternResolver(getClass().getClassLoader()).getResources("classpath*:/fasttillegg*.json")) {
            repository.add(objectMapper.readValue(r.getInputStream(), FasttilleggResource.class));
        }
    }

    @Override
    public Set<String> actions() {
        return Stream.of(PersonalActions.UPDATE_FASTTILLEGG).map(Enum::name).collect(Collectors.toSet());
    }

    @Override
    public void accept(Event<FintLinks> response) {
        log.debug("Handling {} ...", response);
        log.trace("Event data: {}", response.getData());
        try {
            switch (PersonalActions.valueOf(response.getAction())) {
                case UPDATE_FASTTILLEGG:
                    List<FasttilleggResource> data = objectMapper.convertValue(response.getData(), objectMapper.getTypeFactory().constructCollectionType(List.class, FasttilleggResource.class));
                    log.trace("Converted data: {}", data);
                    data.stream().filter(i-> i.getSystemId()==null||i.getSystemId().getIdentifikatorverdi()==null).forEach(i->i.setSystemId(identifikatorFactory.create()));
                    response.setResponseStatus(ResponseStatus.ACCEPTED);
                    response.setData(null);
                    behaviours.forEach(b -> data.forEach(b.acceptPartially(response)));
                    if (response.getResponseStatus() == ResponseStatus.ACCEPTED) {
                        response.setData(new ArrayList<>(data));
                        data.forEach(r -> repository.removeIf(i -> i.getSystemId().getIdentifikatorverdi().equals(r.getSystemId().getIdentifikatorverdi())));
                        repository.addAll(data);
                    }
                    break;
                case GET_ALL_FASTTILLEGG:
                    response.setResponseStatus(ResponseStatus.ACCEPTED);
                    response.setData(new ArrayList<>(repository));
                    break;
                default:
                    response.setStatus(Status.ADAPTER_REJECTED);
                    response.setResponseStatus(ResponseStatus.REJECTED);
                    response.setStatusCode("INVALID_ACTION");
                    response.setMessage("Invalid action");
            }
        } catch (Exception e) {
            log.error("Error!", e);
            response.setResponseStatus(ResponseStatus.ERROR);
            response.setMessage(e.getMessage());
        }
    }

}

File: src/main/kotlin/no/fint/provider/springer/model/FasttilleggRepository.kt
package no.fint.provider.springer.model

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.annotation.PostConstruct
import no.fint.event.model.Event
import no.fint.event.model.Operation
import no.fint.event.model.ResponseStatus
import no.fint.event.model.Status
import no.fint.provider.springer.behaviour.Behaviour
import no.fint.provider.springer.service.IdentifikatorFactory
import no.fint.provider.springer.storage.LonnRepository
import no.fint.provider.springer.storage.Wrapper
import no.novari.fint.model.administrasjon.personal.PersonalActions
import no.novari.fint.model.resource.FintLinks
import no.novari.fint.model.resource.administrasjon.personal.FasttilleggResource
import org.springframework.core.io.support.PathMatchingResourcePatternResolver
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.stereotype.Repository
import java.io.IOException
import java.util.concurrent.ConcurrentLinkedQueue

@Repository
class FasttilleggRepository(
    mongoTemplate: MongoTemplate,
    wrapper: Wrapper,
    private val objectMapper: ObjectMapper,
    private val identifikatorFactory: IdentifikatorFactory,
    private val behaviours: List<Behaviour<FasttilleggResource>>
) : LonnRepository(mongoTemplate, wrapper) {

    private val repository: MutableCollection<FasttilleggResource> = ConcurrentLinkedQueue()

    @PostConstruct
    @Throws(IOException::class)
    fun init() {
        for (r in PathMatchingResourcePatternResolver(javaClass.classLoader).getResources("classpath*:/fasttillegg*.json")) {
            repository.add(objectMapper.readValue(r.inputStream, FasttilleggResource::class.java))
        }
    }

    override fun actions(): Set<String> = setOf(PersonalActions.UPDATE_FASTTILLEGG.name, PersonalActions.GET_ALL_FASTTILLEGG.name)

    override fun accept(response: Event<FintLinks>) {
        try {
            when (PersonalActions.valueOf(response.action)) {
                PersonalActions.UPDATE_FASTTILLEGG -> {
                    val data: List<FasttilleggResource> = objectMapper.convertValue(
                        response.data,
                        objectMapper.typeFactory.constructCollectionType(List::class.java, FasttilleggResource::class.java)
                    )
                    val conflicts = findConflicts(data, FasttilleggResource::class.java)
                    if (conflicts.isNotEmpty() && response.operation == Operation.CREATE) {
                        response.status = Status.ADAPTER_REJECTED
                        response.responseStatus = ResponseStatus.CONFLICT
                        response.data = ArrayList(conflicts)
                        return
                    }
                    data.filter { it.systemId?.identifikatorverdi == null }
                        .forEach { it.systemId = identifikatorFactory.create() }
                    response.responseStatus = ResponseStatus.ACCEPTED
                    response.data = null
                    behaviours.forEach { b -> data.forEach { b.accept(response, it) } }
                    if (response.responseStatus == ResponseStatus.ACCEPTED) {
                        response.data = ArrayList(data)
                        data.forEach { r -> repository.removeIf { i -> i.systemId?.identifikatorverdi == r.systemId?.identifikatorverdi } }
                        repository.addAll(data)
                    }
                }
                PersonalActions.GET_ALL_FASTTILLEGG -> {
                    response.responseStatus = ResponseStatus.ACCEPTED
                    response.data = ArrayList(repository)
                }
                else -> invalid(response)
            }
        } catch (e: Exception) {
            response.responseStatus = ResponseStatus.ERROR
            response.message = e.message
        }
    }

    private fun invalid(response: Event<FintLinks>) {
        response.status = Status.ADAPTER_REJECTED
        response.responseStatus = ResponseStatus.REJECTED
        response.statusCode = "INVALID_ACTION"
        response.message = "Invalid action"
    }
}
